contracts/multicall/GuardedMulticaller2.sol

```diff
@@ -0,0 +1,240 @@
+
+    /**
+     *
+     * @notice Grants DEFAULT_ADMIN_ROLE to the contract creator
+     * @param _owner Owner of the contract
+     * @param _name Name of the contract
+     * @param _version Version of the contract
+     */
+    // solhint-disable-next-line no-unused-vars
+    constructor(address _owner, string memory _name, string memory _version) EIP712(_name, _version) {
+        _grantRole(DEFAULT_ADMIN_ROLE, _owner);
+    }
+
+    /**
+     *
+     * @notice Execute a list of calls. Returned data from calls are ignored.
+     *  The signature must be generated by an address with MULTICALL_SIGNER_ROLE
+     *  The signature must be valid
+     *  The signature must not be expired
+     *  The reference must be unique
+     *  The reference must not be executed before
+     *  The list of calls must not be empty
+     *  The list of calls is executed in order
+     *
+     * @param _multicallSigner Address of an approved signer
+     * @param _reference Reference
+     * @param _calls List of calls
+     * @param _deadline Expiration timestamp
+     * @param _signature Signature of the multicall signer
+     */
+    // slither-disable-start low-level-calls,cyclomatic-complexity
+    // solhint-disable-next-line code-complexity
+    function execute(
+        address _multicallSigner,
+        bytes32 _reference,
+        Call[] calldata _calls,
+        uint256 _deadline,
+        bytes calldata _signature
+    ) external nonReentrant {
+        // solhint-disable-next-line not-rely-on-time
+        if (_deadline < block.timestamp) {
+            revert Expired(_deadline);
+        }
+        if (_reference == 0) {
+            revert InvalidReference(_reference);
+        }
+        if (replayProtection[_reference]) {
+            revert ReusedReference(_reference);
+        }
+        if (_calls.length == 0) {
+            revert EmptyCallArray();
+        }
+        for (uint256 i = 0; i < _calls.length; i++) {
+            if (bytes(_calls[i].functionSignature).length == 0) {
+                revert InvalidFunctionSignature(_calls[i]);
+            }
+            if (_calls[i].target.code.length == 0) {
+                revert NonContractAddress(_calls[i]);
+            }
+        }
+        if (!hasRole(MULTICALL_SIGNER_ROLE, _multicallSigner)) {
+            revert UnauthorizedSigner(_multicallSigner);
+        }
+
+        // Signature validation
+        if (
+            !SignatureChecker.isValidSignatureNow(
+                _multicallSigner,
+                _hashTypedData(_reference, _calls, _deadline),
+                _signature
+            )
+        ) {
+            revert UnauthorizedSignature(_signature);
+        }
+
+        replayProtection[_reference] = true;
+
+        // Multicall
+        for (uint256 i = 0; i < _calls.length; i++) {
+            bytes4 functionSelector = bytes4(keccak256(bytes(_calls[i].functionSignature)));
+            bytes memory callData = abi.encodePacked(functionSelector, _calls[i].data);
```

Comment:
`data` (bytes) and the `functionSelector` (string) are both dynamic types and could cause a collision in `callData`. Granted a real edge case but could have unintended consequences. Might be worth moving to `abi.encode`. Unless I'm missing something.

Reply:
@audityourcontracts I believe `functionSelector` is a static type `bytes4`. Using `encode` would pad it to 32 bytes. The first 4 bytes would still be read as `functionSelector`, and the other 28 bytes would be considered part of the params, which causes the function not to be called correctly. 

Let me know if that makes sense.

Reply:
The params `data`, which is generated off-chain, would have to use `encode` instead of `encodePacked`, though. Even though I believe solidity uses `decode` to unpack the params, and expects the params data to use `encode`. If params data is accidentally encoded with `encodePacked`, the function call would fail instead of behaving incorrectly.

Reply:
Yeah the `bytes4` cast helps here. Will leave it up to your discretion here.

Reply:
For the code to work, it needs to stay as is. The encoding of function calls is [4 byte function selector][ABI encoded (not packed) parameters]